### ISSUE TITLE
Nil dereference panic in generic scheduler

### ISSUE THREAD

**Issue by arjunrn** [kubernetes/kubernetes#89006]
**What happened**:
We see nil dereference panics in the scheduler:
```
runtime.go:78] Observed a panic: \"invalid memory address or nil pointer dereference\" (runtime error: invalid memory address or nil pointer dereference)
{"log":"goroutine 1531102 [running]:
{"log":"k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1811be0, 0x2b39e00)
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
{"log":"k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
{"log":"panic(0x1811be0, 0x2b39e00)
{"log":"\u0009/usr/local/go/src/runtime/panic.go:679 +0x1b2
{"log":"k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).podFitsOnNode(0xc0001bef00, 0xc006232ee0, 0xc002d10800, 0x1d0d240, 0xc00584eab0, 0x0, 0xc0004d7500, 0x1d51080, 0xc0000f2420, 0xc000a00f00, ...)
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:676 +0x1a1
{"log":"k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).findNodesThatFit.func1(0x2f)
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:492 +0x1b1
{"log":"k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1(0xc007d8b560, 0xc007c98c00, 0xc00027c148, 0xc0029081c0)
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:57 +0xbf
{"log":"created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil
{"log":"\u0009/workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:49 +0x148
","stream":"stderr","time":"2020-03-06T13:58:36.366535516Z"}
```

**What you expected to happen**:
Not see panics

**How to reproduce it (as minimally and precisely as possible)**:
Currently unknown

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-13T18:08:14Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7", GitCommit:"be3d344ed06bff7a4fc60656200a93c74f31f9a4", GitTreeState:"clean", BuildDate:"2020-02-11T19:24:46Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: AWS `c5.2xlarge` instance
- OS (e.g: `cat /etc/os-release`):
```NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
```
- Kernel (e.g. `uname -a`): `Linux ip-172-31-3-120 4.15.0-1058-aws #60-Ubuntu SMP Wed Jan 15 22:35:20 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by Huang-Wei** [comment]
From the symptom, it seems the `info` or `info.Node()` is nil. I recall this issue has been fixed combinely by refactoring of the whole snapshot logic, as well as #84014 and #84232. Could you give a try on 1.17?

**Comment by szuecs** [comment]
This happened again, not yet tested in 1.17.

```
E0317 22:19:37.022722       1 factory.go:585] pod is already present in the activeQ
E0317 22:19:37.043665       1 factory.go:585] pod is already present in the activeQ
E0317 22:19:37.080943       1 factory.go:585] pod is already present in the activeQ
E0317 22:19:37.158274       1 factory.go:585] pod is already present in the activeQ
E0317 22:19:37.169216       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.617435       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.702688       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.720206       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.755610       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.768793       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.819505       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:40.852648       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:42.057283       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:42.510919       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:45.009001       1 factory.go:585] pod is already present in the activeQ
E0317 22:20:45.044889       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 68791 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1811be0, 0x2b39e00)
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1811be0, 0x2b39e00)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).podFitsOnNode(0xc0006980c0, 0xc0044651e0, 0xc0028e6400, 0x1d0d240, 0xc004466e10, 0x0, 0xc0001dce10, 0x1d51080, 0xc000336160, 0xc0008e4f00, ...)
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:676 +0x1a1
k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).findNodesThatFit.func1(0x66)
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:492 +0x1b1
k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1(0xc003a367b0, 0xc004045600, 0xc0008f0360, 0xc000321ab0)
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:57 +0xbf
created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil
        /workspace/anago-v1.16.7-beta.0.33+fd5f0b8795e3d6/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/parallelizer.go:49 +0x148
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1613341]
```

**Comment by ahg-g** [comment]
Thanks for reporting it, can you tell us anything about the workload or the cluster that may help us debug this issue?

**Comment by ahg-g** [comment]
As Wei mentioned, this code has been completely refactored, it would be great if you can test on 1.17 and let us know if you see a similar problem.

We should definitely fix this for 1.16, what we need to do is to check in [podFitsOnNode](https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/scheduler/core/generic_scheduler.go#L649) on whether the underlying node actually exist or not. The reason this is not panicing earlier is because all predicates has a check like this: https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/scheduler/algorithm/predicates/predicates.go#L1738

**Comment by ahg-g** [comment]
The problem is that some NodeInfos in the snapshot have a nil Node(), this happens when a node delete event is received before the delete events of all pods on that node.

Aldo changed this behavior in 1.18, and so 1.18 should not have this problem: https://github.com/kubernetes/kubernetes/commit/afe3b907a18a243891a4525e7306880bd32e162f#diff-b63746327d4db70d74d673864973ce04

In 1.15 and before, it seems that all predicates had a check against a nil Node() and returned error, so we have no panics for older versions. But in 1.16 and 1.17, some code related to the framework didn't have this check and so the panic.

I think the simplest thing we can do is just return unschedulableunresolable at the beginning of podFitsOnNode if nodeinfo.Node() is nil for both 1.17 and 1.16. We can do that for the master and port back to 1.18 just in case.
